File: api/routers/wiki.py
from fastapi import APIRouter, HTTPException, Query
from typing import Optional, List
from ..schemas.wiki import FactorInfo, StrategyInfo
from ..services.wiki_service import wiki_service

router = APIRouter()


@router.get("/factors", response_model=List[FactorInfo])
async def get_factors(
    category: Optional[str] = Query(None, description="Filter by category"),
    source: Optional[str] = Query(None, description="Filter by source"),
    sort: str = Query("updated", description="Sort field"),
    limit: int = Query(50, ge=1, le=500),
):
    """Get list of factors with optional filtering"""
    factors = await wiki_service.get_factors(
        category=category,
        source=source,
        sort=sort,
        limit=limit,
    )
    return factors


@router.get("/factors/search")
async def search_factors(
    q: str = Query(..., description="Search query"),
    limit: int = Query(20, ge=1, le=100),
):
    """Search factors"""
    results = await wiki_service.search(query=q, type="factor", limit=limit)
    return results


@router.get("/factors/{name}")
async def get_factor(name: str):
    """Get factor details by name"""
    factor = await wiki_service.get_factor(name)
    if not factor:
        raise HTTPException(status_code=404, detail=f"Factor '{name}' not found")
    return factor


@router.post("/factors", status_code=201)
async def create_factor(factor: FactorInfo):
    """Create a new factor"""
    result = await wiki_service.create_factor(factor.model_dump(exclude_unset=True))
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.put("/factors/{name}")
async def update_factor(name: str, factor: FactorInfo):
    """Update an existing factor"""
    result = await wiki_service.update_factor(name, factor.model_dump(exclude_unset=True))
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.delete("/factors/{name}")
async def delete_factor(name: str):
    """Delete a factor"""
    result = await wiki_service.delete_factor(name)
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.get("/strategies", response_model=List[StrategyInfo])
async def get_strategies(
    category: Optional[str] = Query(None, description="Filter by category"),
    sort: str = Query("updated", description="Sort field"),
    limit: int = Query(50, ge=1, le=500),
):
    """Get list of strategies with optional filtering"""
    strategies = await wiki_service.get_strategies(
        category=category,
        sort=sort,
        limit=limit,
    )
    return strategies


@router.get("/strategies/search")
async def search_strategies(
    q: str = Query(..., description="Search query"),
    limit: int = Query(20, ge=1, le=100),
):
    """Search strategies"""
    results = await wiki_service.search(query=q, type="strategy", limit=limit)
    return results


@router.get("/strategies/{name}")
async def get_strategy(name: str):
    """Get strategy details by name"""
    strategy = await wiki_service.get_strategy(name)
    if not strategy:
        raise HTTPException(status_code=404, detail=f"Strategy '{name}' not found")
    return strategy


@router.post("/strategies", status_code=201)
async def create_strategy(strategy: StrategyInfo):
    """Create a new strategy"""
    result = await wiki_service.create_strategy(strategy.model_dump(exclude_unset=True))
    if "error" in result:
        raise HTTPException(status_code=400, detail=result["error"])
    return result


@router.get("/search")
async def search_wiki(
    q: str = Query(..., description="Search query"),
    type: str = Query("all", description="Search type: all, factor, strategy"),
    limit: int = Query(20, ge=1, le=100),
):
    """Search across all wiki content"""
    results = await wiki_service.search(query=q, type=type, limit=limit)
    return results


@router.get("/status")
async def get_wiki_status():
    """Get wiki system status"""
    return await wiki_service.get_status()

File: frontend/src/stores/wiki.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { wikiApi } from '@/api/wiki'
import type { FactorInfo, StrategyInfo } from '@/api/wiki'

export const useWikiStore = defineStore('wiki', () => {
  const factors = ref<FactorInfo[]>([])
  const strategies = ref<StrategyInfo[]>([])
  const currentFactor = ref<FactorInfo | null>(null)
  const currentStrategy = ref<StrategyInfo | null>(null)
  const isLoading = ref(false)
  const searchQuery = ref('')
  const selectedCategory = ref<string | undefined>(undefined)

  const fetchFactors = async (params?: { category?: string; source?: string; sort?: string; limit?: number }) => {
    isLoading.value = true
    try {
      const data = await wikiApi.getFactors(params)
      factors.value = data
    } catch (error) {
      console.error('Failed to fetch factors:', error)
      factors.value = []
    } finally {
      isLoading.value = false
    }
  }

  const fetchFactor = async (name: string) => {
    isLoading.value = true
    try {
      const data = await wikiApi.getFactor(name)
      currentFactor.value = data
    } catch (error) {
      console.error('Failed to fetch factor:', error)
      currentFactor.value = null
    } finally {
      isLoading.value = false
    }
  }

  const createFactor = async (factor: Partial<FactorInfo>) => {
    try {
      const data = await wikiApi.createFactor(factor)
      await fetchFactors()
      return data
    } catch (error) {
      console.error('Failed to create factor:', error)
      throw error
    }
  }

  const updateFactor = async (name: string, factor: Partial<FactorInfo>) => {
    try {
      const data = await wikiApi.updateFactor(name, factor)
      await fetchFactors()
      return data
    } catch (error) {
      console.error('Failed to update factor:', error)
      throw error
    }
  }

  const deleteFactor = async (name: string) => {
    try {
      await wikiApi.deleteFactor(name)
      await fetchFactors()
    } catch (error) {
      console.error('Failed to delete factor:', error)
      throw error
    }
  }

  const fetchStrategies = async (params?: { category?: string; sort?: string; limit?: number }) => {
    isLoading.value = true
    try {
      const data = await wikiApi.getStrategies(params)
      strategies.value = data
    } catch (error) {
      console.error('Failed to fetch strategies:', error)
      strategies.value = []
    } finally {
      isLoading.value = false
    }
  }

  const fetchStrategy = async (name: string) => {
    isLoading.value = true
    try {
      const data = await wikiApi.getStrategy(name)
      currentStrategy.value = data
    } catch (error) {
      console.error('Failed to fetch strategy:', error)
      currentStrategy.value = null
    } finally {
      isLoading.value = false
    }
  }

  const createStrategy = async (strategy: Partial<StrategyInfo>) => {
    try {
      const data = await wikiApi.createStrategy(strategy)
      await fetchStrategies()
      return data
    } catch (error) {
      console.error('Failed to create strategy:', error)
      throw error
    }
  }

  const updateStrategy = async (name: string, strategy: Partial<StrategyInfo>) => {
    try {
      const data = await wikiApi.updateStrategy(name, strategy)
      await fetchStrategies()
      return data
    } catch (error) {
      console.error('Failed to update strategy:', error)
      throw error
    }
  }

  const deleteStrategy = async (name: string) => {
    try {
      await wikiApi.deleteStrategy(name)
      await fetchStrategies()
    } catch (error) {
      console.error('Failed to delete strategy:', error)
      throw error
    }
  }

  const searchFactors = async (query: string) => {
    isLoading.value = true
    try {
      const data = await wikiApi.search({ q: query, type: 'factor' })
      factors.value = data.map((r: any) => ({
        name: r.page_name?.split('/').pop() || r.name || '',
        formula: r.formula || '',
        source: r.source || r.page_type || '',
        category: r.category || 'other',
        tags: r.tags || [],
        description: r.description || r.content || '',
        ic_mean: r.ic_mean,
        ic_std: r.ic_std,
        icir: r.icir,
        rank_ic_mean: r.rank_ic_mean,
      }))
    } catch (error) {
      console.error('Failed to search factors:', error)
    } finally {
      isLoading.value = false
    }
  }

  const searchStrategies = async (query: string) => {
    isLoading.value = true
    try {
      const data = await wikiApi.search({ q: query, type: 'strategy' })
      strategies.value = data.map((r: any) => ({
        name: r.page_name?.split('/').pop() || r.name || '',
        description: r.description || r.content || '',
        category: r.category || 'general',
        tags: r.tags || [],
        strategy_yaml: r.strategy_yaml || '',
      }))
    } catch (error) {
      console.error('Failed to search strategies:', error)
    } finally {
      isLoading.value = false
    }
  }

  return {
    factors,
    strategies,
    currentFactor,
    currentStrategy,
    isLoading,
    searchQuery,
    selectedCategory,
    fetchFactors,
    fetchFactor,
    createFactor,
    updateFactor,
    deleteFactor,
    fetchStrategies,
    fetchStrategy,
    createStrategy,
    updateStrategy,
    deleteStrategy,
    searchFactors,
    searchStrategies,
  }
})
